### PR TITLE
Record: optimise filterWithIndex (conservative) and traverseWithKey (less object creations) (fixes #711)

### LINKED ISSUE #711: `insert` of `Record` should be a `noop` if the value at the key already exists (strict equals)
Note: We may also optimise `filterWithIndex` and `filterMapWithIndex`

### PR DESCRIPTION
**Pull request by sledorze**

### CODE REVIEW

**Comment by sledorze** [comment]
@gcanti I still have the problem of running the tests locally (related to Jest 
 as depicted here https://github.com/gcanti/fp-ts/pull/715
).
So blindly guessing and using the CI to fix issues, sorry for the noise..

**Comment by gcanti** [comment]
> I still have the problem of running the tests locally

Ah! Adding

```json
"jest": {
  "testURL": "http://localhost/"
}
```

in `package.json` does solve the problem?

**Comment by gcanti** [comment]
@sledorze actually we could upgrade to `jest@23.6.0`

File: src/Record.ts [hunk 1]
```diff
@@ -571,15 +580,25 @@ export function filterMapWithIndex<A, B>(
 export function filterWithIndex<K extends string, A>(fa: Record<K, A>, p: (key: K, a: A) => boolean): Record<string, A>
 export function filterWithIndex<A>(fa: Record<string, A>, p: (key: string, a: A) => boolean): Record<string, A>
 export function filterWithIndex<A>(fa: Record<string, A>, p: (key: string, a: A) => boolean): Record<string, A> {
```
Comment (gcanti) [inline src/Record.ts]: What about

```ts
export function filterWithIndex<A>(fa: Record<string, A>, p: (key: string, a: A) => boolean): Record<string, A> {
  const r: Record<string, A> = {}
  let changed = false
  for (const key in fa) {
    if (fa.hasOwnProperty(key)) {
      const a = fa[key]
      if (p(key, a)) {
        r[key] = a
      } else {
        changed = true
      }
    }
  }
  return changed ? r : fa
}
```

File: test/Record.ts [hunk 2]
```diff
@@ -147,6 +149,12 @@ describe('Record', () => {
     const y: Record<string, string | number> = { a: 1, b: 'foo' }
     const actual = R.filter(y, isNumber)
     assert.deepStrictEqual(actual, { a: 1 })
+
+    const z: Record<string, string | number> = { b: 1, a: 'foo', c: 'bar' }
```
Comment (gcanti) [inline test/Record.ts]: This test doesn't look different from the previous one

**Comment by sledorze** [comment]
Was to trigger another case for coverage

File: src/Record.ts (hunk 1 above)
Comment (sledorze) [reply src/Record.ts]: It's doing all work in case of noop (obj création plus assigns (n log n))

Comment (gcanti) [reply src/Record.ts]: noop is unlikely in a `filter*` function so in the most common case looks like you are doing even more work

Comment (sledorze) [reply src/Record.ts]: I m fine with the simpler approach.
Afk this morning (on phone) so expect a change later.

**Comment by sledorze** [comment]
@gcanti yes adding

```
"jest": {
  "testURL": "http://localhost/"
}
```
Solves the problem (I have another one popping, however)

**Comment by gcanti** [comment]
FYI we are upgrading jest in the other PR https://github.com/gcanti/fp-ts/pull/715#issuecomment-462247271

**Comment by sledorze** [comment]
@gcanti not sure how I can fix that coverage issue

**Comment by gcanti** [comment]
@sledorze did you try `open coverage/lcov-report/index.html`?

File: src/Record.ts [hunk 3]
```diff
@@ -572,14 +581,18 @@ export function filterWithIndex<K extends string, A>(fa: Record<K, A>, p: (key:
...
   const r: Record<string, A> = {}
-  const keys = Object.keys(fa)
-  for (const key of keys) {
-    const a = fa[key]
-    if (p(key, a)) {
-      r[key] = a
+  let changed = false
+  for (const key in fa) {
+    if (fa.hasOwnProperty(key)) {
```
Comment (gcanti) [inline src/Record.ts]: ah sorry, do you mean this line right?

**Comment by gcanti** [comment]
this seems to work

```ts
const x = Object.assign(Object.create({ c: true }), { a: 1, b: 'foo' })
assert.deepStrictEqual(R.filter(x, isNumber), { a: 1 })
```

File: src/Record.ts (hunk 3 above)
Comment (sledorze) [reply src/Record.ts]: yes

**Comment by sledorze** [comment]
thanks @gcanti

**Pull request merged**